commit 438623a06bacd69c40c4af633bb09a3bbb9dfc78
Author: Chuck Lever <chuck.lever@oracle.com>
Date:   Thu Jul 15 15:52:06 2021 -0400

    SUNRPC: Add svc_rqst::rq_auth_stat
    
    I'd like to take commit 4532608d71c8 ("SUNRPC: Clean up generic
    dispatcher code") even further by using only private local SVC
    dispatchers for all kernel RPC services. This change would enable
    the removal of the logic that switches between
    svc_generic_dispatch() and a service's private dispatcher, and
    simplify the invocation of the service's pc_release method
    so that humans can visually verify that it is always invoked
    properly.
    
    All that will come later.
    
    First, let's provide a better way to return authentication errors
    from SVC dispatcher functions. Instead of overloading the dispatch
    method's *statp argument, add a field to struct svc_rqst that can
    hold an error value.
    
    Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
    Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>

diff --git a/net/sunrpc/svcauth.c b/net/sunrpc/svcauth.c
index 998b196b6176..5a8b8e03fdd4 100644
--- a/net/sunrpc/svcauth.c
+++ b/net/sunrpc/svcauth.c
@@ -59,12 +59,12 @@ svc_put_auth_ops(struct auth_ops *aops)
 }
 
 int
-svc_authenticate(struct svc_rqst *rqstp, __be32 *authp)
+svc_authenticate(struct svc_rqst *rqstp)
 {
 	rpc_authflavor_t	flavor;
 	struct auth_ops		*aops;
 
-	*authp = rpc_auth_ok;
+	rqstp->rq_auth_stat = rpc_auth_ok;
 
 	flavor = svc_getnl(&rqstp->rq_arg.head[0]);
 
@@ -72,7 +72,7 @@ svc_authenticate(struct svc_rqst *rqstp, __be32 *authp)
 
 	aops = svc_get_auth_ops(flavor);
 	if (aops == NULL) {
-		*authp = rpc_autherr_badcred;
+		rqstp->rq_auth_stat = rpc_autherr_badcred;
 		return SVC_DENIED;
 	}
 
@@ -80,7 +80,7 @@ svc_authenticate(struct svc_rqst *rqstp, __be32 *authp)
 	init_svc_cred(&rqstp->rq_cred);
 
 	rqstp->rq_authop = aops;
-	return aops->accept(rqstp, authp);
+	return aops->accept(rqstp);
 }
 EXPORT_SYMBOL_GPL(svc_authenticate);